src/support.cpp

```diff
@@ -48,6 +48,322 @@ bool AreaSupport::handleSupportModifierMesh(SliceDataStorage& storage, const Set
+            support_infill_part.infill_wall = support_infill_part.insets[0];
+            // the area for generating infill patterns may or may not be there
+            if (wall_line_count == support_infill_part.insets.size())
+            {
+                support_infill_part.infill_area = support_infill_part.insets[wall_line_count - 1];
+            }
+
+            storage.support.supportLayers[layer_nr].support_infill_parts.push_back(support_infill_part);
+        }
+    }
+
+    // compute different density areas for each support island
+    for (unsigned int layer_nr = 0; layer_nr < total_layer_count - 1; ++layer_nr)
+    {
+        if (layer_nr < min_layer || layer_nr > max_layer)
+        {
+            continue;
+        }
+
+        // generate separate support islands and calculate density areas for each island
+        std::vector<SupportInfillPart>& support_infill_parts = storage.support.supportLayers[layer_nr].support_infill_parts;
+        for (unsigned int part_idx = 0; part_idx < support_infill_parts.size(); ++part_idx)
+        {
+            SupportInfillPart& support_infill_part = support_infill_parts[part_idx];
+            AABB thisPartBoundaryBox(support_infill_part.infill_area);
+
+            // calculate density areas for this island
+            Polygons less_dense_support = support_infill_part.infill_area; // one step less dense with each density_step
+            for (unsigned int density_step = 0; density_step < max_density_steps; ++density_step)
+            {
+                size_t min_layer = layer_nr + density_step * gradual_support_step_layer_count + layer_skip_count;
+                size_t max_layer = layer_nr + (density_step + 1) * gradual_support_step_layer_count;
+
+                for (float upper_layer_idx = min_layer; static_cast<unsigned int>(upper_layer_idx) <= max_layer; upper_layer_idx += layer_skip_count)
+                {
+                    if (static_cast<unsigned int>(upper_layer_idx) >= total_layer_count)
+                    {
+                        less_dense_support.clear();
+                        break;
+                    }
+
+                    // compute intersections with relevent upper parts
+                    const std::vector<SupportInfillPart> upper_infill_parts = storage.support.supportLayers[upper_layer_idx].support_infill_parts;
+                    Polygons relevent_upper_polygons;
+                    for (unsigned int upper_part_idx = 0; upper_part_idx < upper_infill_parts.size(); ++upper_part_idx)
+                    {
+                        // we compute intersection based on support infill areas
+                        AABB upperPartBoundaryBox(upper_infill_parts[upper_part_idx].infill_area);
+
+                        if (upperPartBoundaryBox.hit(thisPartBoundaryBox))
+                        {
+                            relevent_upper_polygons.add(upper_infill_parts[upper_part_idx].infill_area);
```

Comment:
```
legend:
^ support roof
| support wall
# dense support
+ less dense support

comparing infill    comparing with outline
   ^^^^^^           ^^^^^^
   ####||^^         ####||^^
   ######||^^       ######||^^
   ++++####||       ++++++##||
   ++++++####       ++++++++##
```

I'm not a 100% sure, but I think the second method is better.
Whichever you use, please document clearly why you chose that way of implementing it.
Please **explain** your choice to me and clearly **document** the code with your choice.